californium-core/src/test/java/org/eclipse/californium/core/network/UdpMatcherTest.java

```diff
@@ -62,6 +65,28 @@ public void testReceiveResponseAcceptsResponseWithoutCorrelationInformation() {
 	}
 
 	@Test
+	public void testCancelledRequestShouldRemoveFromTokenMap() {
+		UdpMatcher matcher = newMatcher(false);
+		Exchange exchange = sendRequest(matcher, null);
+
+		exchange.getRequest().cancel();
+
+		Exchange matched = matcher.receiveResponse(responseFor(exchange.getCurrentRequest()), null);
+		assertNull(matched);
+	}
+
+	@Test
+	public void testCancelledRequestShouldRemoveFromMidMap() {
```

Comment:
Any reason to not just blindly remove cancelled requests? I understand that the leak just happens when a response for acknowledged request is lost. But since we introduced a cleanup-on-cancel mechanism, is there any advantage to being selective on clean up? Seems safe to just clean up everything if the user cancelled the request. Also creates a consistent user experience, where cancelled requests invoke error callback, regardless of whether they received an ACK or not.